### ISSUE TITLE
Form key length limit 2048 exceeded

### ISSUE THREAD

**Issue by sadjadbp** [dotnet/aspnetcore#18871]
I have a similar issue to #11262 , where I'm randomly getting "Form key length limit 2048 exceeded" after upgrading to asp.net core 3.1 from asp.net core 2.2

Situation is I'm passing a large Base64Url encoded to client and getting it back in a POST call. Then I'll read the posted field using Request.Form["fieldname"]

I'm getting "Form key length limit 2048 exceeded". At first I've started to add the attribute `        [RequestFormLimits(KeyLengthLimit = 30_000, ValueCountLimit = 1500, ValueLengthLimit = 100_000)]` to be able to get to the actual problem, which is that form parser seems to be getting parts of the value of that field as the Form Key.

To explain better let's have an example:

Form contains:
Arg1: Value1 (small string),
Arg2: Value2 (very large base64url string like eyJUYXJnZXRCYXNlTmFtZSI6IjkzM2U0Yzc2...)
Arg3: Value3,
Arg4: Value4
...
ArgN: ValueN

When it reaches the Request.Form["Arg2"] it's empty! 

Now I've tried to dump the whole Request.Form

```csharp
foreach (var formFieldName in httpContext.Request.Form.Keys.Take(100))
```

And I see in my logs that form looks like
Arg1 key: Value1
Arg2 parts of Value2 like nZXRCYX [not the Arg2 key, but part of value become the key]: empty value
Arg3 key: some times Value3, some times rest of Arg2 Value
Arg4 key: some time Value4, some times rest of Arg2 Value
...
ArgN: ValueN

that was why I was getting that exception "Form key length limit 2048 exceeded" since parts of value was part of being parsed as the key! which make sense.

Now also I've added some code to be able to dump the whole Request.Body and it's okay. Also I've tried to call again using that same Body and nothing happens. Everyday we have 3-4 requests failing like this. So somehow this new FormPipeReader is flaky with Request.Form.

**Comment by javiercn** [comment]
@sadjadbp thanks for contacting us.

@Tratcher this belongs in hosting, doesn't it?

**Comment by Tratcher** [comment]
Can you share a Fiddler trace so we can see the raw request body?

**Comment by sadjadbp** [comment]
[form.zip](https://github.com/dotnet/aspnetcore/files/4172869/form.zip)

Attached a saz file for my testing in localhost (so it won't contain any re-usable secrets).
As explained above I cannot repro this and nor does our clients, it just happens randomly, but I have been dumping all the details for the failed requests and couldn't find anything special about them (unicode vs non-unicode chars or anything like that), and as mentioned if I send the same failing body of request it would pass locally for me.

Explanation of the saz file:
The first GET request to is the where we are generating the form content with "ViewState" field as the Base64Url encoded value (json with some encrypted info in it)

The rest is trying some commands where each would cause a POST. 

Where our customers are hitting error is that on some POSTs the ViewState field is missing (the key ViewState is not there and the value [or parts it] is being parsed as the key) to give you an example here is the dump of Request.Forms in asp.net core (I have redacted some values with xxxxx so it won't contain any secrets)
```
HttpContext.Request.Forms:
Key:submitArgs, Value:NextPage
Key:N4NGw4bDJPexxxxxZTkyYmJkNThjIn0, Value:
Key:_CER_HIDDEN, Value:HOW_MANY_TIMES_MARRIED,HOW_MANY_TIMES_MARRIED
Key:PET_ADDRPOSTALCODE, Value:
Key:PET_RESFROM, Value:06/01%2N0STFRZ05hN2JGaXNqNXpYMGx0V05zazFrL1ZRNWg3a2g2KzNEb3QrUmNJcEhKTGpUWVZJRFpSODNiN2E5ZjlUQ01Zdytxxxxx MHNLamEwU3FMdmc4dHY4T0NpTE1rRnFwYVVxxxxx N2pJbTVuMkcyZTYrTjJWdmIn0,06/01/2012
Key:_CER_FOOTER, Value:
Key:_CER_DATAINTEGRITY, Value:28db9bbf840e4e1cb976020e92bbd58c
Key:__RequestVerificationToken, Value:CfDJ8ODqPmxOEjdCvTrO0mwUwnfSw-wfY4QpCF5xHsMdhI6rY-mR6PovmGJ-PdCbK9_3PdQsrCmuZlRroB6g6gCGrtCetYWd7JUKjCS6zQrVaDebPCa2F563qUyzwO4GKnlbu5GyW6-qdYNidFJiQNm8kuQenQYzCFiaIP4HsitVAOanJbCICQvqTsQip2BWYEYcEA
```

As you see "ViewState" key is missing, but as you see the second key is part of ViewState value and value is empty. Also as you see"PET_RESFROM" value has the rest of ViewState as value; rest is all okay. 

Also I have added this piece of code to dump the request body

```csharp
var httpContext = HttpContextAccessor?.HttpContext;
if ((httpContext?.Request?.Body?.CanRead ?? false)
    && (httpContext.Request.Method == HttpMethods.Post || httpContext.Request.Method == HttpMethods.Put))
{
    httpContext.Request.EnableBuffering(); // internally would call BufferingHelper.EnableRewind(request);
    httpContext.Request.Body.Seek(0, SeekOrigin.Begin);
    using (var reader = new StreamReader(httpContext.Request.Body, leaveOpen: true))
    {
        string requestBody = await reader.ReadToEndAsync();

        var chuncks = requestBody.Chunks(8000).ToList(); // AppInsights max size
        for (int i = 0; i < chuncks.Count; i++)
        {
            props.Add($"HttpContext.Request.Body_{i}", chuncks[i]);
        }
    }
    httpContext.Request.Body.Seek(0, SeekOrigin.Begin); // making sure it's back to the original state (seems like this would throw)
}
```
and when I look at the body it looks very normal and I've tried to resend the same body in my local with no repro.

```
submitArgs=NextPage&ViewState=eyJUYXJnZXRCYXNlTmFtxxxxxxiOTc2MDIwZTkyYmJkNThjIn0&_CER_HIDDEN=HOW_MANY_TIMES_MARRIED&PET_RESFROM_CAL=06%2F01%2F2012&PET_PADDRSTR1_CAL=&PET_PADDRAPT1_CAL=&PET_PADDRCITY1_CAL=&_CER_NOTECOPY=&_CER_TACOPY=&_CER_NOTEMODS=&_CER_TAMODS=&_CER_FIELDMODS=&_CER_FOOTER=&_CER_DATAINTEGRITY=28db9bbf840e4e1cb976020e92bbd58c&__RequestVerificationToken=CfDJ8ODqPmxOEjdCvTrO0mwUwnfSw-wfY4QpCF5xHsMdhI6rY-mR6PovmGJ-PdCbK9_3PdQsrCmuZlRroB6g6gCGrtCetYWd7JUKjCS6zQrVaDebPCa2F563qUyzwO4GKnlbu5GyW6-qdYNidFJiQNm8kuQenQYzCFiaIP4HsitVAOanJbCICQvqTsQip2BWYEYcEA
```

This is what body looks like

**Comment by Tratcher** [comment]
You're right, the forms look fine, there does seem to be an inconsistent issue with the parser. When does your buffering code capture the body? In a middleware or in your controller? Capturing it after MVC had read the form wouldn't work unless the form reader had independently enabled buffering.

**Comment by sadjadbp** [comment]
it's not in middleware, it's in MVC.

I have decorated my POST method with `ReadableBodyStream` which is defined as

```csharp
public class ReadableBodyStreamAttribute : AuthorizeAttribute, IAuthorizationFilter
{
    public void OnAuthorization(AuthorizationFilterContext context)
    {
        context.HttpContext.Request.EnableBuffering();
    }
}
```

When I check and see the Request.Forms["ViewState"] is empty, I'm dumping all the Request,Forms and Request.Body as explained above.

**Comment by Tratcher** [comment]
Hmm, the buffering should have made you take the fast path in the form parser which is the most consistent and tested. We're going to have to experiment until we repro this, even if inconstantly. Maybe using a body stream that returns one byte at a time to check for boundary conditions.

**Comment by sadjadbp** [comment]
@Tratcher is there a way for us to switch back to FormReader (from FormPipeReader) so we can mitigate this for now and also be sure FormPipeReader is the culprit?

**Comment by Tratcher** [comment]
The old form parser is still [there](https://github.com/dotnet/aspnetcore/blob/master/src/Http/WebUtilities/src/FormReader.cs), you can call it from somewhere like ReadableBodyStreamAttribute and set the result to HttpRequest.Form.

**Comment by anurse** [comment]
Triage: It looks like a read-across-segment-boundary issue. @Tratcher will take a look at the provided test data and force some boundaries and see what we get.

**Comment by Tratcher** [comment]
I've done some experiments with single byte segments, single byte reads, the file buffering read stream, etc. and I have not been able to reproduce any boundary condition related corruption. Putting this on hold until someone can come up with another theory on how to repro.

**Comment by Tratcher** [comment]
Proposal: Fuzz testing by returning segments of random sizes.
